### ISSUE TITLE
`COPY tbl TO 'output.parquet' (FORMAT PARQUET);` outputs incomplete statistics in metadata with complex types if fields are `NULL`

### ISSUE THREAD

**Issue by mitchellciupak** [duckdb/duckdb#12958]
# What happens?

## Overview
I am attempting to export tables as parquet files with the `COPY tbl TO 'output.parquet' (FORMAT PARQUET);` syntax from the [Parquet Export](https://duckdb.org/docs/guides/file_formats/parquet_export.html) docs and make use of `pyiceberg.Table().add_files()` method for registering parquet files as datafiles to an [Apache Iceberg](https://iceberg.apache.org) [Table](https://py.iceberg.apache.org/reference/pyiceberg/table/) like seen [here](https://github.com/apache/iceberg-python/blob/main/mkdocs/docs/api.md#add-files), but the [`pyiceberg.table.Table.add_files()` method](https://github.com/apache/iceberg-python/blob/main/pyiceberg/table/__init__.py#L1641) fails to validates the existance of statistics in the parquet metadata in a function called [`data_file_statistics_from_parquet_metadata`](https://github.com/apache/iceberg-python/blob/main/pyiceberg/io/pyarrow.py#L1838), assumedly per the [Apache Iceberg Spec](https://iceberg.apache.org/spec/).

[`pyiceberg.io.pyarrow.data_file_statistics_from_parquet_metadata()`](https://github.com/apache/iceberg-python/blob/main/pyiceberg/io/pyarrow.py#L1838) attempts the validation using [a member of the `pyarrow.parquet.ColumnChunkMetaData` class called `is_stats_set: bool`]((https://arrow.apache.org/docs/python/generated/pyarrow.parquet.ColumnChunkMetaData.html#pyarrow.parquet.ColumnChunkMetaData.is_stats_set)) from [`pyarrow's latest version (currently v16.1.0)`](https://github.com/apache/iceberg-python/blob/main/pyproject.toml#L61) as seen [on this line of the pyiceberg src](https://github.com/apache/iceberg-python/blob/main/pyiceberg/io/pyarrow.py#L1910). For a simpler look, so we can all now forget about `pyiceberg`, here is a snippet to express the validation that is failing in isolation:
```python
import pyarrow.parquet as pq # version 16.1.0

parquet_file: str = "/path/to/file.parquet"

parquet_metadata = pq.read_metadata(parquet_file)

for r in range(parquet_metadata.num_row_groups):
    row_group = parquet_metadata.row_group(r)

    for pos in range(parquet_metadata.num_columns):
        column = row_group.column(pos)
        if column.is_stats_set:
            continue
        else:
            print(f"PyArrow statistics missing for column {pos}!")
```
If I am reading the [pyarrow source code correctly for `is_stats_set()`](https://github.com/apache/arrow/blob/main/cpp/src/parquet/metadata.cc#L254C15-L254C29), then I assume per [this `MakeStatistics()` function call](https://github.com/apache/arrow/blob/main/cpp/src/parquet/metadata.cc#L93) and [this `MakeStatistics()` function definition](https://github.com/apache/arrow/blob/main/cpp/src/parquet/statistics.h#L361), that `null_count`, `distinct_count`, `num_values`, `min_value`, and `max_value` need to be set in the parquet file's metadata in order to pass this check.

After creating a few types of tables, exporting them to parquet, and checking their metadata with [`parquet_metadata()`](https://duckdb.org/docs/data/parquet/metadata#parquet-metadata), I found that these statistics some of these are missing from columns with an `ARRAY`, `LIST`, `MAP`, or `STRUCT` type when the table contains rows in which those columns are null.

## TLDR
After creating a few types of tables, exporting them to parquet, and checking their metadata with [`parquet_metadata()`](https://duckdb.org/docs/data/parquet/metadata#parquet-metadata), I found that these statistics some of these are missing from columns with an `ARRAY`, `LIST`, `MAP`, or `STRUCT` type when the table contains rows in which those columns are null.

## Notes
* [location in arrow src of `is_stats_set()`](https://github.com/apache/arrow/blob/main/cpp/src/parquet/metadata.cc#L254)
* [`pyiceberg.table.Table.add_files()`](https://github.com/apache/iceberg-python/blob/main/pyiceberg/table/__init__.py#L1641)'s call sequence to [`pyiceberg.io.pyarrow.data_file_statistics_from_parquet_metadata()`](https://github.com/apache/iceberg-python/blob/main/pyiceberg/io/pyarrow.py#L1838) goes:
    1. [`pyiceberg.table.Table.add_files()`](https://github.com/apache/iceberg-python/blob/main/pyiceberg/table/__init__.py#L1641)
    2. [`pyiceberg.table.Transaction.add_files()`](https://github.com/apache/iceberg-python/blob/main/pyiceberg/table/__init__.py#L686C9-L686C18)
    3. [`pyiceberg.table.Table._parquet_files_to_data_files()`](https://github.com/apache/iceberg-python/blob/main/pyiceberg/table/__init__.py#L3077)
    4. [`pyiceberg.io.pyarrow.parquet_files_to_data_files()`](https://github.com/apache/iceberg-python/blob/main/pyiceberg/io/pyarrow.py#L2035)
    5. [`pyiceberg.io.pyarrow.data_file_statistics_from_parquet_metadata()`](https://github.com/apache/iceberg-python/blob/main/pyiceberg/io/pyarrow.py#L1838)
* I removed the `INTERVAL` type, listed in [SQL Data Types](https://duckdb.org/docs/sql/data_types/overview#nested--composite-types), from my testing after finding that no statistics would be created for those columns whatsoever whether the values were null or not. For reference I used values like `INTERVAL '0' SECOND`, `INTERVAL '1' HOUR`, `INTERVAL '100' YEAR`, and `NULL` in ``INSERT INTO tbl VALUES (...);` queries and found that the would be represented as expected in `SELECT * FROM tbl;` queries. I am happy to submit an issue addressing this, although it I assume likeley is more a question of how to represent those values in stats and I really don't have any personal application for it's functionality.
* I also removed the `BIT` type, listed in [SQL Data Types](https://duckdb.org/docs/sql/data_types/overview#nested--composite-types), from my testing after finding that `COPY tbl TO 'output.parquet' (FORMAT PARQUET);` raises a `Not implemented Error: Unimplemented type for Parquet "BIT"` when the column type is present in the table. Again, more than happy to submit an issue for it if that would be helpful.

# To Reproduce

### Install and Start
1. `mkdir expmt-duckdb-parquet-export-tbl-w-complex-types && cd expmt-duckdb-parquet-export-tbl-w-complex-types` 
2. Installation per [DuckDB Installation](https://duckdb.org/docs/installation/?version=stable&environment=cli&platform=macos&download_method=direct) docs, `wget -q https://github.com/duckdb/duckdb/releases/download/v1.0.0/duckdb_cli-osx-universal.zip -O duckdb_cli-osx-universal.zip && unzip -q duckdb_cli-osx-universal.zip && mv duckdb duckdb_v1.0.0 && rm duckdb_cli-osx-universal.zip`
3. `./duckdb_v1.0.0`
### Test Parquet Export without Nulls
1. CREATE TABLE (in duckdb shell) with all supported SQL data types, per [Data Types](https://duckdb.org/docs/sql/data_types/overview) docs
```sql
CREATE TABLE tbl_without_nulls (
    bigint_int8_long_type BIGINT,
    blob_bytea_binary_varbinary_type BLOB,
    boolean_bool_logical_type BOOLEAN,
    date_type DATE,
    decimal_numeric_type DECIMAL(18,3),
    double_float8_type DOUBLE,
    hugeint_type HUGEINT,
    integer_int4_int_signed_type INTEGER,
    real_float4_float_type REAL,
    smallint_int2_short_type SMALLINT,
    time_type TIME,
    timestamptz_type TIMESTAMPTZ,
    timestamp_datetime_type TIMESTAMP,
    tinyint_int1_type TINYINT,
    ubigint_type UBIGINT,
	  uhugeint_type UHUGEINT,
	  uinteger_type UINTEGER,
	  usmallint_type USMALLINT,
	  utinyint_type UTINYINT,
	  uuid_type UUID,
    varchar_char_bpchar_text_string_type VARCHAR,
    array_integer_type INTEGER[3],
    list_integer_type INTEGER[],
    map_integer_varchar_type MAP(INTEGER, VARCHAR),
    struct_integer_varchar_type STRUCT(i INTEGER, j VARCHAR),
    union_integer_varchar_type UNION(num INTEGER, text VARCHAR)
);
```
2. NSERT INTO (in duckdb shell) non-null values, somewhat per [Numeric Types](https://duckdb.org/docs/sql/data_types/numeric.html) Docs
```sql
-- Insert min values
INSERT INTO tbl_without_nulls VALUES
(
    -9223372036854775808,  -- min BIGINT
    x'00',                 -- min BLOB
    FALSE,                 -- min BOOLEAN
    '0001-01-01',          -- min DATE
    -999999999999.999,     -- min DECIMAL(18,3)
    -1.7976931348623157E+308,  -- min DOUBLE
    -170141183460469231731687303715884105728, -- min HUGEINT
    -2147483648,           -- min INTEGER
    -3.4028235E+38,        -- min REAL
    -32768,                -- min SMALLINT
    '00:00:00',            -- min TIME
    '0001-01-01 00:00:00+00',  -- min TIMESTAMPTZ
    '0001-01-01 00:00:00',     -- min TIMESTAMP
    -128,                  -- min TINYINT
    0,                     -- min UBIGINT
    0,                     -- min UHUGEINT
    0,                     -- min UINTEGER
    0,                     -- min USMALLINT
    0,                     -- min UTINYINT
    '00000000-0000-0000-0000-000000000000',  -- min UUID
    '',                    -- min VARCHAR
    ARRAY[0, 0, 0],        -- min INTEGER[3]
    ARRAY[],               -- min INTEGER[]
    MAP([0, 1], ['a', 'b']),
    {'i': 0, 'j': 'a'},
    union_value(num := 0)
);

-- Insert middle-ish values
INSERT INTO tbl_without_nulls VALUES
(
    0,                     -- middle BIGINT
    x'01',                 -- middle BLOB
    TRUE,                  -- middle BOOLEAN
    '2024-01-01',          -- middle DATE
    0.000,                 -- middle DECIMAL(18,3)
    0.0,                   -- middle DOUBLE
    0,                     -- middle HUGEINT
    0,                     -- middle INTEGER
    0.0,                   -- middle REAL
    0,                     -- middle SMALLINT
    '12:00:00',            -- middle TIME
    '2024-01-01 12:00:00+00',  -- middle TIMESTAMPTZ
    '2024-01-01 12:00:00',     -- middle TIMESTAMP
    0,                     -- middle TINYINT
    9223372036854775807,   -- middle UBIGINT (since it's unsigned)
    170141183460469231731687303715884105727, -- middle UHUGEINT
    2147483647,            -- middle UINTEGER
    65535,                 -- middle USMALLINT
    255,                   -- middle UTINYINT
    '123e4567-e89b-12d3-a456-426614174000',  -- middle UUID
    'Middle Text',         -- middle VARCHAR
    ARRAY[1, 2, 3],        -- middle INTEGER[3]
    ARRAY[1, 2, 3, 4],     -- middle INTEGER[]
    MAP([1, 2], ['a', 'b']),
    {'i': 42, 'j': 'a'},
    union_value(num := 2)
);

-- Insert max values
INSERT INTO tbl_without_nulls VALUES
(
    9223372036854775807,   -- max BIGINT
    x'FF',                 -- max BLOB (just assuming 1-byte length)
    TRUE,                  -- max BOOLEAN
    '9999-12-31',          -- max DATE
    999999999999.999,      -- max DECIMAL(18,3)
    1.7976931348623157E+308, -- max DOUBLE
    170141183460469231731687303715884105727, -- max HUGEINT
    2147483647,            -- max INTEGER
    3.4028235E+38,         -- max REAL
    32767,                 -- max SMALLINT
    '23:59:59',            -- max TIME
    '9999-12-31 23:59:59+00', -- max TIMESTAMPTZ
    '9999-12-31 23:59:59',    -- max TIMESTAMP
    127,                   -- max TINYINT
    18446744073709551615,  -- max UBIGINT
    34028236692093846346337, -- not-max UHUGEINT
    4294967295,            -- max UINTEGER
    65535,                 -- max USMALLINT
    255,                   -- max UTINYINT
    'ffffffff-ffff-ffff-ffff-ffffffffffff', -- max UUID
    'Maximim Text',            -- max VARCHAR
    ARRAY[2147483647, 2147483647, 2147483647], -- max INTEGER[3]
    ARRAY[2147483647, 2147483647, 2147483647, 2147483647], -- max INTEGER[]
    MAP([2147483647, 2147483646], ['max1', 'max2']),
    {'i': 2147483647, 'j': 'Max Text'},
    union_value(num := 3)
);
```
3. `SELECT * FROM tbl_without_nulls;` just to make sure they are there
4. `COPY tbl_without_nulls TO 'tbl_without_nulls.duckdb_v1.0.0.parquet' (FORMAT PARQUET);` ,per [Parquet Export](https://duckdb.org/docs/guides/file_formats/parquet_export.html) Docs
5. `SELECT column_id, type, stats_min, stats_max, stats_null_count, stats_distinct_count, stats_min_value, stats_max_value FROM parquet_metadata('tbl_without_nulls.duckdb_v1.0.0.parquet');`
6. `python pyarrow_read_metadata.py tbl_without_nulls.duckdb_v1.0.0.parque` (the python snippet mentioned in `What happens?`) ⇒ no output expected

### Test Parquet Export with Nulls
1. CREATE TABLE (in duckdb shell) with all supported SQL data types, per [Data Types](https://duckdb.org/docs/sql/data_types/overview) docs
```sql
CREATE TABLE tbl_without_nulls (
    bigint_int8_long_type BIGINT,
    blob_bytea_binary_varbinary_type BLOB,
    boolean_bool_logical_type BOOLEAN,
    date_type DATE,
    decimal_numeric_type DECIMAL(18,3),
    double_float8_type DOUBLE,
    hugeint_type HUGEINT,
    integer_int4_int_signed_type INTEGER,
    real_float4_float_type REAL,
    smallint_int2_short_type SMALLINT,
    time_type TIME,
    timestamptz_type TIMESTAMPTZ,
    timestamp_datetime_type TIMESTAMP,
    tinyint_int1_type TINYINT,
    ubigint_type UBIGINT,
	  uhugeint_type UHUGEINT,
	  uinteger_type UINTEGER,
	  usmallint_type USMALLINT,
	  utinyint_type UTINYINT,
	  uuid_type UUID,
    varchar_char_bpchar_text_string_type VARCHAR,
    array_integer_type INTEGER[3],
    list_integer_type INTEGER[],
    map_integer_varchar_type MAP(INTEGER, VARCHAR),
    struct_integer_varchar_type STRUCT(i INTEGER, j VARCHAR),
    union_integer_varchar_type UNION(num INTEGER, text VARCHAR)
);
```
2. INSERT INTO (in duckdb shell) null values, somewhat per [Numeric Types](https://duckdb.org/docs/sql/data_types/numeric.html) Docs
```sql
-- Insert Null Values
INSERT INTO tbl_with_nulls VALUES
(
    NULL,
    NULL,
    NULL,
    NULL,
    NULL,
    NULL,
    NULL,
    NULL,
    NULL,
    NULL,
    NULL,
    NULL,
    NULL,
    NULL,
    NULL,
    NULL,
    NULL,
    NULL,
    NULL,
    NULL,
    NULL,
    NULL,
    NULL,
    NULL,
    NULL,
    NULL,
);
INSERT INTO tbl_with_nulls VALUES
(
    NULL,
    NULL,
    NULL,
    NULL,
    NULL,
    NULL,
    NULL,
    NULL,
    NULL,
    NULL,
    NULL,
    NULL,
    NULL,
    NULL,
    NULL,
    NULL,
    NULL,
    NULL,
    NULL,
    NULL,
    NULL,
    NULL,
    NULL,
    NULL,
    NULL,
    NULL,
);
INSERT INTO tbl_with_nulls VALUES
(
    NULL,
    NULL,
    NULL,
    NULL,
    NULL,
    NULL,
    NULL,
    NULL,
    NULL,
    NULL,
    NULL,
    NULL,
    NULL,
    NULL,
    NULL,
    NULL,
    NULL,
    NULL,
    NULL,
    NULL,
    NULL,
    NULL,
    NULL,
    NULL,
    NULL,
    NULL,
);
```
3. `SELECT * FROM tbl_with_nulls;` just to make sure they are there
4. `COPY tbl_with_nulls TO 'tbl_with_nulls.duckdb_v1.0.0.parquet' (FORMAT PARQUET);` ,per [Parquet Export](https://duckdb.org/docs/guides/file_formats/parquet_export.html) Docs
5. `SELECT column_id, type, stats_min, stats_max, stats_null_count, stats_distinct_count, stats_min_value, stats_max_value FROM parquet_metadata('tbl_with_nulls.duckdb_v1.0.0.parquet');`
6. `python pyarrow_read_metadata.py tbl_with_nulls.duckdb_v1.0.0.parquet` (the python snippet mentioned in `What happens?`)
    1. expected to output:
        ```
        PyArrow statistics missing for column 21!
        PyArrow statistics missing for column 22!
        PyArrow statistics missing for column 23!
        PyArrow statistics missing for column 24!
        ```
    2. which correlates to:
        ```
        array_integer_type INTEGER[3],
        list_integer_type INTEGER[],
        map_integer_varchar_type MAP(INTEGER, VARCHAR),
        struct_integer_varchar_type STRUCT(i INTEGER, j VARCHAR),
       ```
## Also tested within the following environments and received the same results:
* cli, duckdb nightly v1.0.1-dev2472, aarch64 MacOS v14.5
* python api, duckdb v1.0.0, python v3.12.3, aarch64 MacOS v14.5
* python api, duckdb v1.0.0, python v3.11.8, aarch64 MacOS v14.5
* python api, duckdb v1.0.0, python:3.11.8-slim container via rancher-desktop, aarch64 MacOS v14.5
* cli, duckdb v1.0.0, x86 Windows 10 Enterprise Version 22H2 OS build 19045.4529
* cli, duckdb nightly v1.0.1-dev2472, x86 Windows 10 Enterprise Version 22H2 OS build 19045.4529

## OS:

aarch64 MacOS v14.5 via rancher-desktop

## DuckDB Version:

1.0.0, but also tested on nighty v1.0.1-dev2472

## DuckDB Client:

CLI, but also tested via Python API

## Full Name:

Mitchell Ciupak

## Affiliation:

Comcast

## What is the latest build you tested with? If possible, we recommend testing with the latest nightly build.

I have tested with a nightly build

## Did you include all relevant data sets for reproducing the issue?

Yes

## Did you include all code required to reproduce the issue?

- [X] Yes, I have

## Did you include all relevant configuration (e.g., CPU architecture, Python version, Linux distribution) to reproduce the issue?

- [X] Yes, I have

**Comment by hannes** [comment]
Hey, thanks for this detailed report. Not sure who is in the wrong here. I've tried reproducing the issue that you described, we indeed do not set statistics on complex types if *all* values are NULL because what would the stats value be. If a complex type contains *some* actual values, the stats seem to be correctly set. In general, we don't set the `disctinct_count` and `stats_null_count` statistic because they're non-trivial to compute. Could you maybe try to boil this down a little more?

Here's what I tried (in the CLI):

```SQL
create table test(l integer[]);
insert into test values (null), (null), (null);
copy (from test) to '/tmp/test.parquet';
.null NULL
select path_in_schema, columns('stats*') from parquet_metadata('/tmp/test.parquet');
```

```
┌──────────────────┬───────────┬───────────┬──────────────────┬──────────────────────┬─────────────────┬─────────────────┐
│  path_in_schema  │ stats_min │ stats_max │ stats_null_count │ stats_distinct_count │ stats_min_value │ stats_max_value │
│     varchar      │  varchar  │  varchar  │      int64       │        int64         │     varchar     │     varchar     │
├──────────────────┼───────────┼───────────┼──────────────────┼──────────────────────┼─────────────────┼─────────────────┤
│ l, list, element │ NULL      │ NULL      │             NULL │                 NULL │ NULL            │ NULL            │
└──────────────────┴───────────┴───────────┴──────────────────┴──────────────────────┴─────────────────┴─────────────────┘
```

We can see that indeed all the stats are NULL. Now lets see what happens if I insert some data:

```SQL
insert into test values ([42, 42, 43]);
copy (from test) to '/tmp/test.parquet';
select path_in_schema, columns('stats*') from parquet_metadata('/tmp/test.parquet');
```

```
┌──────────────────┬───────────┬───────────┬──────────────────┬──────────────────────┬─────────────────┬─────────────────┐
│  path_in_schema  │ stats_min │ stats_max │ stats_null_count │ stats_distinct_count │ stats_min_value │ stats_max_value │
│     varchar      │  varchar  │  varchar  │      int64       │        int64         │     varchar     │     varchar     │
├──────────────────┼───────────┼───────────┼──────────────────┼──────────────────────┼─────────────────┼─────────────────┤
│ l, list, element │ 42        │ 43        │             NULL │                 NULL │ 42              │ 43              │
└──────────────────┴───────────┴───────────┴──────────────────┴──────────────────────┴─────────────────┴─────────────────┘
```

This looks all pretty good to me.

**Comment by mitchellciupak** [comment]
Hi @hannes, thanks for taking a look. To your point at the beginning, I'm not sure if either tool is wrong, but I have have been able to generate statistics for columns containing null values by either exporting a duckdb `.execute()` to a `pyarrow.Table` using `.arrow()` or exporting a parquet file from duckdb and rewriting the metadata with [`pyarrow.parquet.write_metadata`](https://arrow.apache.org/docs/python/generated/pyarrow.parquet.write_metadata.html) similar to how [`parquet.write_metadata`](pyarrow.parquet.write_metadata) does when it's `write_statistics` parameter is set to `True`. Would it be helpful to reproduce its output in comparison to the `.parquet` files created by duckdb. This may be a feature request to pass a similar parameter to the aforementioned `write_statistics`

**Comment by mitchellciupak** [comment]
@hannes, I think note would be that statistics are still missing for complex types when even ~~some~~ values are null (1 or more). Would that be helpful to reproduce?

**Comment by hannes** [comment]
Even, if *some* values are NULL, the stats are still there:

```SQL
create table test(l integer[]);
insert into test values (null), ([42, 42, 43]), (null);
copy (from test) to '/tmp/test.parquet';
select path_in_schema, columns('stats*') from parquet_metadata('/tmp/test.parquet');
```

```
┌──────────────────┬───────────┬───┬─────────────────┬─────────────────┐
│  path_in_schema  │ stats_min │ … │ stats_min_value │ stats_max_value │
│     varchar      │  varchar  │   │     varchar     │     varchar     │
├──────────────────┼───────────┼───┼─────────────────┼─────────────────┤
│ l, list, element │ 42        │ … │ 42              │ 43              │
├──────────────────┴───────────┴───┴─────────────────┴─────────────────┤
│ 1 rows                                           7 columns (4 shown) │
└──────────────────────────────────────────────────────────────────────┘
```

**Comment by mitchellciupak** [comment]
```sql
create table test(l integer[]);
insert into test values ([null]), ([42, 42, 43]), (null);
copy (from test) to '/tmp/test.parquet';
select path_in_schema, columns('stats*') from parquet_metadata('/tmp/test.parquet');
```
```
┌──────────────────┬───────────┬───────────┬──────────────────┬──────────────────────┬─────────────────┬─────────────────┐
│  path_in_schema  │ stats_min │ stats_max │ stats_null_count │ stats_distinct_count │ stats_min_value │ stats_max_value │
│     varchar      │  varchar  │  varchar  │      int64       │        int64         │     varchar     │     varchar     │
├──────────────────┼───────────┼───────────┼──────────────────┼──────────────────────┼─────────────────┼─────────────────┤
│ l, list, element │ 42        │ 43        │                  │                      │ 42              │ 43              │
└──────────────────┴───────────┴───────────┴──────────────────┴──────────────────────┴─────────────────┴─────────────────┘
```
Not `null_count` or `distinct_count`

**Comment by mitchellciupak** [comment]
```python
# python3 -m venv ./venv && source ./venv/bin/activate
# pip install pyarrow
# python pyarrow_rewriter_footer.py

import os
import io
import datetime
from collections import defaultdict

import pyarrow as pa
import pyarrow.parquet as pq

parquet_file: str = "/tmp/test.parquet"

def get_min_value(current_min, new_min):
    """Safely compares and returns the minimum of two values, handling None values and type mismatches."""
    if current_min is None:
        return new_min
    if new_min is None:
        return current_min

    # Ensure both are of the same type for comparison
    if isinstance(current_min, str) and isinstance(new_min, str):
        return min(current_min, new_min)
    if isinstance(current_min, (int, float)) and isinstance(new_min, (int, float)):
        return min(current_min, new_min)
    if isinstance(current_min, (pa.TimestampType, datetime.datetime)) and isinstance(new_min, (pa.TimestampType, datetime.datetime)):
        return min(current_min, new_min)
    # If types are different, skip comparison and return the current min (customize as needed)
    return current_min

def get_max_value(current_max, new_max):
    """Safely compares and returns the maximum of two values, handling None values and type mismatches."""
    if current_max is None:
        return new_max
    if new_max is None:
        return current_max

    # Ensure both are of the same type for comparison
    if isinstance(current_max, str) and isinstance(new_max, str):
        return max(current_max, new_max)
    if isinstance(current_max, (int, float)) and isinstance(new_max, (int, float)):
        return max(current_max, new_max)
    if isinstance(current_max, (pa.TimestampType, datetime.datetime)) and isinstance(new_max, (pa.TimestampType, datetime.datetime)):
        return max(current_max, new_max)
    # If types are different, skip comparison and return the current max (customize as needed)
    return current_max

def rewrite_parquet_footer(file_path: str, output_path: str) -> None:
    # Check if the file exists and is not empty
    if not os.path.isfile(file_path) or os.path.getsize(file_path) < 8:
        raise ValueError(f"Parquet file {file_path} is too small or does not exist.")

    # Open the input file and read the metadata
    parquet_metadata = pq.read_metadata(file_path)

    # Create a ParquetFile object for reading row groups
    parquet_file = pq.ParquetFile(file_path)

    # Initialize a dictionary to store updated column-level statistics
    updated_statistics = defaultdict(lambda: {'min': None, 'max': None, 'null_count': 0, 'total_values': 0})

    # Iterate over row groups to gather new statistics for each column
    for row_group_index in range(parquet_metadata.num_row_groups):
        row_group = parquet_metadata.row_group(row_group_index)
        for column_index in range(row_group.num_columns):
            column_chunk = row_group.column(column_index)
            column_name = parquet_metadata.schema.column(column_index).path[0]

            # Set statistics from each row group
            updated_statistics[column_name]['total_values'] = (
                updated_statistics[column_name].get('total_values', 0) + (column_chunk.num_values or 0)
            )
            if column_chunk.statistics is not None:
                updated_statistics[column_name]['null_count'] = (
                    updated_statistics[column_name].get('null_count', 0) + (column_chunk.statistics.null_count or 0)
                )


            if 'total_values' not in updated_statistics[column_name] and updated_statistics[column_name]['total_values'] is not None:
                updated_statistics[column_name]['total_values'] = column_chunk.num_values or 0
            else:
                updated_statistics[column_name]['total_values'] += column_chunk.num_values or 0

            if column_chunk.statistics is not None:
                if 'null_count' not in updated_statistics[column_name] and updated_statistics[column_name]['null_count'] is not None:
                    updated_statistics[column_name]['null_count'] = column_chunk.statistics.null_count or 0 
                else:
                    updated_statistics[column_name]['null_count'] += column_chunk.statistics.null_count or 0

            # Update min and max if present
            try:
                if column_chunk.statistics is not None and column_chunk.statistics.has_min_max:
                    min_value = column_chunk.statistics.min
                    max_value = column_chunk.statistics.max

                    # Use a helper function to perform type-safe comparison
                    updated_statistics[column_name]['min'] = get_min_value(
                        updated_statistics[column_name]['min'], min_value
                    )
                    updated_statistics[column_name]['max'] = get_max_value(
                        updated_statistics[column_name]['max'], max_value
                    )
            except Exception as e:
                print(f"Error processing column {column_name} in row group {row_group_index}: {e}")
                continue

    # Ensure statistics have sensible default values and are never None
    for column_stats in updated_statistics.values():
        column_stats['min'] = column_stats['min'] if column_stats['min'] is not None else b''
        column_stats['max'] = column_stats['max'] if column_stats['max'] is not None else b''
        column_stats['null_count'] = column_stats['null_count'] if column_stats['null_count'] is not None else 0
        column_stats['total_values'] = column_stats['total_values'] if column_stats['total_values'] is not None else 0

    # Create a new Parquet file with the updated statistics as metadata
    with pq.ParquetWriter(output_path, parquet_metadata.schema.to_arrow_schema(), compression='gzip') as writer:
        # Open the input file again to copy row groups
        for row_group_index in range(parquet_metadata.num_row_groups):
            # Read each row group using ParquetFile.read_row_group
            row_group_table = parquet_file.read_row_group(row_group_index)
            writer.write_table(row_group_table)
            

    print(f"Successfully rewritten Parquet file {output_path} with updated statistics.")


rewrite_parquet_footer(parquet_file, "/tmp/test.rewrite.parquet")
```
if the stats are recalculate and added to the footer stats using pyarrow with the script above, then
```sql
select path_in_schema, columns('stats*') from parquet_metadata('/tmp/test.rewrite.parquet');
┌──────────────────┬───────────┬───────────┬──────────────────┬──────────────────────┬─────────────────┬─────────────────┐
│  path_in_schema  │ stats_min │ stats_max │ stats_null_count │ stats_distinct_count │ stats_min_value │ stats_max_value │
│     varchar      │  varchar  │  varchar  │      int64       │        int64         │     varchar     │     varchar     │
├──────────────────┼───────────┼───────────┼──────────────────┼──────────────────────┼─────────────────┼─────────────────┤
│ l, list, element │ 42        │ 43        │                2 │                      │ 42              │ 43              │
└──────────────────┴───────────┴───────────┴──────────────────┴──────────────────────┴─────────────────┴─────────────────┘
```
Would It be possible to include a calculation for `stats_null_count` in the parquet export?

**Comment by mitchellciupak** [comment]
Hey @hannes or @szarnyasg, I was wondering if this issue has seen any changes in status since placed `under review`

**Comment by mitchellciupak** [comment]
^ @hannes or @szarnyasg?

**Comment by samansmink** [comment]
Hey @mitchellciupak I took another look and indeed we still do not write the null count, i've put this in our internal todos to look into, but can not make any promises related to when this would get resolved
